### PR TITLE
Improve managed identity support

### LINKED ISSUE #5779: Improve managed identity support
We want to support managed identity in App Service, Cloud Shell, and generic compute VMs (via IMDS). `msrestazure` has more robust implementations of MSI and IMDS auth based on ADAL. That code can't be copied over but is a good guide to error handling and endpoint behavior, including some apparently undocumented cases.

### PR DESCRIPTION
**Pull request by chlowell**
This refactors `MsiCredential` and `ImdsCredential` to support the three managed identity environments we've identified, and user-assigned identities:
- General VM (`ImdsCredential`)
  - Probing the endpoint is the only way to determine whether it's available. Currently we send an HTTP request with a short timeout. Trying to establish a TCP connection would be neater but would break the current tests.
  - GET static IMDS endpoint with header `Metadata: true`
  - client id of user-assigned identity is a query param
- App Service (`MsiCredential`)
  - `MSI_ENDPOINT` and `MSI_SECRET` are set
  - GET `MSI_ENDPOINT` with header `secret: $MSI_SECRET`
  - client id of user-assigned identity is a query param
- Cloud Shell (`MsiCredential`)
  - `MSI_ENDPOINT` alone is set
  - POST `MSI_ENDPOINT` with header `Metadata: true`
  - client id of user-assigned identity is form data

Both credentials follow the [retry guidance for the IMDS endpoint](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#retry-guidance).

This also adds some synchronous tests for managed identity scenarios, adapted from `msrestazure`. These use [`httpretty`](https://pypi.org/project/httpretty/) to validate requests and mock responses. There are no equivalent async tests yet because I can't get `httpretty` to work properly with async code. I'll probably rework these tests to use standard library `mock` instead or, failing that, some other library with async support (tracked by #5899).

Closes #5779

### CODE REVIEW

**Comment by adxsdk6** [comment]
Can one of the admins verify this patch?

File: sdk/identity/azure-identity/azure/identity/constants.py [hunk 1]
```diff
@@ -2,7 +2,9 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License. See LICENSE.txt in the project root for
 # license information.
-# --------------------------------------------------------------------------
+# -------------------------------------------------------------------------
+MSI_ENDPOINT = "MSI_ENDPOINT"
+MSI_SECRET = "MSI_SECRET"
```
Comment (schaabs) [inline sdk/identity/azure-identity/azure/identity/constants.py]: Out of curiosity why didn't you add this to the EnvironmentVariables class as with the others?

File: sdk/identity/azure-identity/azure/identity/aio/_internal.py [hunk 2]
```diff
@@ -4,85 +4,102 @@
...
+            # we send a request it would immediately reject (missing a required header),
+            # setting a short timeout.
+            try:
+                await self._client.request_token(scopes, method="GET", connection_timeout=0.3)
+                self._endpoint_available = True
+            except AuthenticationError:
+                # received a response that couldn't be deserialized (because it was an error response)
+                self._endpoint_available = True
+            except ConnectTimeoutError:
+                self._endpoint_available = False
```
Comment (schaabs) [inline sdk/identity/azure-identity/azure/identity/aio/_internal.py]: Are there other possible error types besides AuthenticationError and ConnectTimeoutError possible here?  Should we catch more broadly for the case when we set _enpoint_available to False?

File: sdk/identity/azure-identity/azure/identity/constants.py (hunk 1 above)
Comment (chlowell) [reply sdk/identity/azure-identity/azure/identity/constants.py]: Good question. I think past me had a reason for this that's irrelevant now so I've moved it into the class.

File: sdk/identity/azure-identity/azure/identity/aio/_internal.py (hunk 2 above)
Comment (chlowell) [reply sdk/identity/azure-identity/azure/identity/aio/_internal.py]: Changed this to consider HttpResponseError (got a response, some policy choked on it) and AuthenticationError (got a response, AuthnClient choked on it) as indicating availability, anything else as indicating unavailability.

**Pull request merged**